### PR TITLE
Migrate-withdraw-rewards

### PR DESCRIPTION
**Pull request by wolverineks**
migrates withdraw staking rewards to yoroi-lib
migrates transaction type to `YoroiUnsignedTx` and `YoroiSignedTx`
migrates to hooks

this might break voting registration, staking, and sending

### CODE REVIEW

File: src/Dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -316,3 +301,30 @@ const useBalances = (_wallet: YoroiWallet) => {
     {},
   )
 }
+
+const txHistoryRoute = {
```
Comment (wolverineks) [inline src/Dashboard/Dashboard.tsx]: `const { resetToTxHistory } = useWalletNavigation()`

File: polyfills.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+if (!Object.entries) {
```
Comment (wolverineks) [inline polyfills.ts]: `Object.fromEntries()`

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 3]
```diff
@@ -541,17 +535,26 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
     }
   }
 
-  async signTx(signRequest: UnsignedTx, decryptedMasterKey: string) {
+  async signTx(unsignedTx: YoroiUnsignedTx, decryptedMasterKey: string) {
     const masterKey = await Bip32PrivateKey.fromBytes(Buffer.from(decryptedMasterKey, 'hex'))
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:539]: move this stuff to yoroi-lib?

File: src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  )
+
+  return (
+    <>
+      <TwoActionView
+        title={strings.confirmTx}
+        primaryButton={{
+          disabled: isLoading,
+          label: strings.confirmButton,
+          onPress: async () =>
```
Comment (stackchain) [inline src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx]:
```suggestion
          onPress: () =>
```

File: src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            ensureKeysValidity(wallet.id)
+              .then(() =>
+                navigation.navigate('biometrics', {
+                  keyId: wallet.id,
+                  onSuccess: async (masterKey) => {
```
Comment (stackchain) [inline src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx]:
```suggestion
                  onSuccess: (masterKey) => {
```

File: src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithHW.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  onCancel: () => void
+  onSuccess: () => void
+}
+
+export const ConfirmTxWithHW: React.FC<Props> = ({wallet, unsignedTx, onSuccess, onCancel}) => {
+  const strings = useStrings()
+  const [transport, setTransport] = React.useState<'USB' | 'BLE'>('USB')
+  const [step, setStep] = React.useState<'select-transport' | 'connect-transport' | 'confirm'>('select-transport')
+
+  const onSelectTransport = async (transport: 'USB' | 'BLE') => {
```
Comment (stackchain) [inline src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithHW.tsx]:
```suggestion
  const onSelectTransport = (transport: 'USB' | 'BLE') => {
```

File: src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+      <LoadingOverlay loading={isLoading} />
+    </>
+  )
+}
+
+const LoadingOverlay: React.FC<{loading: boolean}> = ({loading}) => {
+  return loading ? (
+    <View style={StyleSheet.absoluteFill}>
+      <View style={[StyleSheet.absoluteFill, {opacity: 0.5, backgroundColor: 'pink'}]} />
```
Comment (stackchain) [inline src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx:78]: I think we need an example on Figma, not sure about `pink`.

File: src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx [hunk 8]
```diff
@@ -1,356 +1,166 @@
...
+        <Route active={true}>
+          <ConfirmTx
+            wallet={wallet}
+            storage={storage}
+            unsignedTx={state.withdrawalTx}
+            onSuccess={onSuccess}
+            onCancel={onCancel}
+          />
+        </Route>
+      )}
```
Comment (stackchain) [inline src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx:46]: Won't work?
```suggestion
      <Route active={state.step === 'confirm'}>
        <ConfirmTx
          wallet={wallet}
          storage={storage}
          unsignedTx={state.withdrawalTx}
          onSuccess={onSuccess}
          onCancel={onCancel}
        />
      </Route>
```

**Review by stackchain — APPROVED** [review]

File: src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx (hunk 7 above)
Comment (wolverineks) [reply src/Dashboard/WithdrawStakingRewards/ConfirmTx/ConfirmTxWithOS.tsx:78]: yeah, that was for testing...

File: src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx (hunk 8 above)
Comment (wolverineks) [reply src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx:46]: when step === 'form'
withdrawalTx === undefined

when step === 'confirm'
withdrawalTx === YoroiUnsignedTx

**Pull request merged**